docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,149 @@
+  This function will be called by `oecore` just prior to the termination of enclave thread.
+  Thread-specific exit functions (`__cxa_thread_atexit` functions) will be executed prior to `oe_allocator_thread_cleanup`. </br>
+  This allows the thread-specific exit functions to use memory management functions (e.g to free resources) before the allocator itself is cleaned up. </br>
+  It is recommended that allocators be judicious about using C++11 thread-specific objects in `oe_allocator_thread_cleanup` since
+  use of C++11 thread-local objects may trigger a call to (`__cxa_thread_atexit`) after all the thread-specific exit functions have executed.
+
+
+  It is important that the allocators be aware of the nature of enclave threads, and be designed accordingly. Enclave threads can be short-lived. For OE SGX, enclave threads last only for the duration of an ecall. </br>
+  Additionally, the same thread-control structure (e.g., `sgx_tcs_t` for SGX) can be bound to different host threads at different times during the lifespan of an enclave.
+
+
+- `void* oe_allocator_malloc(size_t size)` </br>
+  This function will be called by `oecore` to implement `malloc`. Depending upon the build settings, the allocated memory might be tracked by the `debug-malloc` feature. `oe_allocator_malloc` must provide the same semantics as [malloc](https://en.cppreference.com/w/c/memory/malloc).
+
+
+- `void oe_allocator_free(void* ptr)` </br>
+  This function will be called by `oecore` to implement `free`. `oe_allocator_free` must provide the same semantics as [free](https://en.cppreference.com/w/c/memory/free).
+
+
+- `void* oe_allocator_calloc(size_t nmemb, size_t size)` </br>
+  This function will be called by `oecore` to implement `calloc`. `oe_allocator_calloc` must provide the same semantics as [calloc](https://en.cppreference.com/w/c/memory/calloc).
+
+
+- `void* oe_allocator_realloc(void* ptr, size_t size)` </br>
+  This function will be called by `oecore` to implement `realloc`. `oe_allocator_realloc` must provide the same semantics as [realloc](https://en.cppreference.com/w/c/memory/realloc).
+
+- `void* oe_allocator_aligned_alloc(size_t alignment, size_t size)` </br>
+  This function will be called by `oecore` to implement [aligned_alloc](https://linux.die.net/man/3/aligned_alloc/man/3/posix_memalign). Memory allocated by this function can be freed via `free`.
+
+- `void oe_allocator_posix_memalign(void** memptr, size_t alignment, size_t size)` </br>
+  This function will be called by `oecore` to implement [posix_memalign](https://linux.die.net/man/3/posix_memalign). Memory allocated by this function can be freed via `free`.
+
+- `size_t oe_allocator_malloc_usable_size(void* ptr)` </br>
+  This function is called by `oecore` to implement [malloc_usable_size](http://man7.org/linux/man-pages/man3/malloc_usable_size.3.html). </br>
+  This function must be implemented. However, allocators may choose to provide an empty implementation.
```

Comment:
It can't really be empty since it has to return a value.  Maybe you mean "an implementation that always returns 0"?  Or what?   Simplest way to resolve is to delete the sentence about empty.
```suggestion
  This function must be implemented.
```

Reply:
Good point. Thanks for the suggestion.